File: content/exercises/intro-python/basics/conditionals/checker.py
import io
import subprocess
import sys


def main() -> None:
    submission_path = sys.argv[1]
    output_path = sys.argv[2]
    tests: list[tuple[str, bool, str]] = []
    feedback: list[str] = []
    status = "failed"

    process = subprocess.run(
        [sys.executable, submission_path],
        capture_output=True,
        text=True,
        check=False,
    )
    if process.returncode != 0:
        tests.append(("code runs", False, "Code did not run successfully."))
        feedback.append(f"Execution error: {process.stderr.strip() or 'runtime error'}")
    else:
        tests.append(("code runs", True, "Code executed without errors."))
        output_lines = [line.strip() for line in process.stdout.splitlines() if line.strip()]
        if "Middle income" in output_lines:
            tests.append(
                ("conditional output", True, "Output includes 'Middle income' for gdp=45000.")
            )
        else:
            tests.append(
                (
                    "conditional output",
                    False,
                    "Expected output 'Middle income' for gdp = 45000.",
                )
            )

    passed_all = tests and all(item[1] for item in tests)
    if passed_all:
        status = "passed"
        feedback.append("Correct solution submitted.")
    elif not feedback:
        feedback.append("Use if/elif/else so gdp = 45000 prints 'Middle income'.")

    output_stream = io.StringIO()
    output_stream.write(f"STATUS:{status}\n")
    for item in feedback:
        output_stream.write(f"FEEDBACK:{item}\n")
    for name, passed, message in tests:
        output_stream.write(f"TEST:{name}|{'pass' if passed else 'fail'}|{message}\n")

    with open(output_path, "w", encoding="utf-8") as handle:
        handle.write(output_stream.getvalue())


if __name__ == "__main__":
    main()

File: content/exercises/intro-r/vectors/vector-ops/checker.R
args <- commandArgs(trailingOnly = TRUE)

if (length(args) < 2) {
  stop("Usage: checker.R <submission_path> <output_path>")
}

submission_path <- args[[1]]
output_path <- args[[2]]

test_results <- list()
feedback <- c()
status <- "failed"

record_test <- function(name, passed, message) {
  test_results[[length(test_results) + 1]] <<- list(
    name = name,
    passed = passed,
    message = message
  )
}

env <- new.env(parent = baseenv())
execution_ok <- TRUE
execution_error <- NULL
captured_output <- character()

tryCatch(
  {
    captured_output <- capture.output(source(submission_path, local = env))
  },
  error = function(e) {
    execution_ok <<- FALSE
    execution_error <<- conditionMessage(e)
  }
)

if (!execution_ok) {
  record_test("code runs", FALSE, "Your code did not run successfully.")
  feedback <- c(feedback, paste("Execution error:", execution_error))
} else {
  record_test("code runs", TRUE, "Code executed without errors.")
  if (!exists("prices", envir = env, inherits = FALSE)) {
    record_test("prices exists", FALSE, "Define vector `prices`.")
  } else {
    prices <- get("prices", envir = env, inherits = FALSE)
    expected <- c(10.5, 20.0, 15.75, 8.25)
    if (is.numeric(prices) && identical(as.numeric(prices), expected)) {
      record_test("prices values", TRUE, "prices vector has expected values.")
    } else {
      record_test("prices values", FALSE, "Set prices to c(10.5, 20.0, 15.75, 8.25).")
    }
  }

  output_text <- paste(captured_output, collapse = "\n")
  has_mean <- grepl("13\\.125", output_text)
  has_sum <- grepl("54\\.5", output_text)
  has_filtered <- grepl("20", output_text) && grepl("15\\.75", output_text)
  record_test("prints mean", has_mean, "Output should include mean(prices) = 13.125.")
  record_test("prints sum", has_sum, "Output should include sum(prices) = 54.5.")
  record_test("prints filtered prices", has_filtered, "Output should include prices above 12.")
}

passed_all <- length(test_results) > 0 && all(vapply(test_results, function(t) isTRUE(t$passed), logical(1)))
if (passed_all) {
  status <- "passed"
  feedback <- c(feedback, "Correct solution submitted.")
} else if (length(feedback) == 0) {
  feedback <- c(feedback, "Compute mean/sum and print prices filtered above 12.")
}

output_lines <- c(paste0("STATUS:", status))
for (msg in feedback) {
  output_lines <- c(output_lines, paste0("FEEDBACK:", msg))
}
for (t in test_results) {
  pass_label <- if (isTRUE(t$passed)) "pass" else "fail"
  output_lines <- c(output_lines, paste0("TEST:", t$name, "|", pass_label, "|", t$message))
}

writeLines(output_lines, con = output_path)
